Support labeling of volumes to Pod's SELinux label

swagiaal opened the pull request:
This is the implementation of #14192 
- Uses docker's :Z option to handle relabling of mounted volumes to be accessible by the containers' SELinux label
- Adds a new function to volume.Builder to enable volumes to express whether relabeling is desired or not.

Note: This depends on #14705 

@pmorie

Commit:
Support volume relabling for pods which specify an SELinux label

pmorie commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -573,13 +573,21 @@ func makeEnvList(envs []kubecontainer.EnvVar) (result []string) {
+		// If the pod does not provide an SELinux context
+		// relabeling will restrict access to the volume the
+		// container which mounts it last
+		if m.SELinuxRelabel && podHasSELinuxLabel {
```
We only want to do this for the first container that mounts the volume, so we need to have the pod and the containers passed into this method.  We should only relabel each volume a single time.

pmorie replied:
Also, if the pod uses the hostPid namespace, we will need to chcon the volume in the docker manager instead of relying on docker.  I think we should extract the `ChconRunner` interface into its own package.  I'm going to add siblings to it.  Following the pattern of mount, I was thinking: `pkg/chmod/chmod.go`:

```go
package chmod

type Interface interface {
  Chmod(path, mode string) error
}
```

Thoughts about that?

swagiaal replied:
> I think we should extract the ChconRunner interface into its own package.

How about 'pkg/selinux/selinux.go' ? we can add all the SELinux utils there

swagiaal replied:
> We only want to do this for the first container that mounts the volume, so we need to have the pod and the containers passed into this method. We should only relabel each volume a single time.

Implemented here: https://github.com/kubernetes/kubernetes/commit/042f3a4920bcee730fb84a13b494b5d82a9cc09c

pmorie commented on pkg/kubelet/kubelet.go:
```diff
@@ -927,10 +927,11 @@ func makeMounts(container *api.Container, podVolumes kubecontainer.VolumeMap) (m
 			continue
 		}
 		mounts = append(mounts, kubecontainer.Mount{
```
This is the right place to calculate whether this is the first container that mounts the volume and drive `SELinuxRelabel` based on that.

pmorie commented on pkg/volume/volume.go:
```diff
@@ -45,6 +45,12 @@ type Builder interface {
+	// volume to match the pod to which it will be attached.
+	// Usually this returns true of the volume type supports
+	// SELinux and the volume is newly formatted.
+	RequiresSELinuxRelabel() bool
```
This is called `SupportsSELinux` in the proposal, I'd prefer to stick with the names we agreed to (even though I don't think they're great) and we can bikeshed around this once the functionality is working.  There's a lot of cleanup to be done on volumes after this, and we'll make it just-so then.

pmorie commented:
@swagiaal Here's a test we can do that doesn't rely on modifying any images:

1.  Make a pod with MCS label `c0,c1` that makes an emptyDir volume
2.  Make a pod with MCS label `c2,c3` that runs on the host the pod in (1) is scheduled onto and does a hostpath mount of the volume directory for that pod.
3.  Verify that pod (2) cannot read or write pod (1)'s volume

Thoughts?

pmorie commented on pkg/volume/empty_dir/empty_dir.go:
```diff
@@ -189,6 +189,10 @@ func (ed *emptyDir) IsReadOnly() bool {
 }
 
+func (ed *emptyDir) RequiresSELinuxRelabel() bool {
+	return false
```
This should be `true`, same with secrets, git repo, downward API volume, etc.

pmorie commented on pkg/volume/fc/fc.go:
```diff
@@ -187,6 +187,10 @@ func (b *fcDiskBuilder) IsReadOnly() bool {
 }
 
+func (b *fcDiskBuilder) RequiresSELinuxRelabel() bool {
+	return false
```
This is a block device, so theoretically this can be true.  @rootfs, thoughts?

pmorie commented on pkg/volume/gce_pd/gce_pd.go:
```diff
@@ -234,6 +234,10 @@ func (b *gcePersistentDiskBuilder) IsReadOnly() bool {
 }
 
+func (b *gcePersistentDiskBuilder) RequiresSELinuxRelabel() bool {
+	return false
```
Should be true.

pmorie commented on pkg/volume/git_repo/git_repo.go:
```diff
@@ -118,6 +118,10 @@ func (b *gitRepoVolumeBuilder) IsReadOnly() bool {
 }
 
+func (b *gitRepoVolumeBuilder) RequiresSELinuxRelabel() bool {
+	return false
```
True

pmorie commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -179,6 +179,10 @@ func (b *iscsiDiskBuilder) IsReadOnly() bool {
 }
 
+func (b *iscsiDiskBuilder) RequiresSELinuxRelabel() bool {
+	return false
```
True, theoretically, but I'm not sure we need to test it.  @childsb, @erinboyd, @wattsteve ?

pmorie commented on pkg/volume/rbd/rbd.go:
```diff
@@ -225,6 +225,10 @@ func (b *rbd) IsReadOnly() bool {
 }
 
+func (b *rbd) RequiresSELinuxRelabel() bool {
+	return false
```
True

pmorie commented on pkg/volume/secret/secret.go:
```diff
@@ -172,6 +172,10 @@ func (sv *secretVolume) IsReadOnly() bool {
 }
 
+func (sv *secretVolume) RequiresSELinuxRelabel() bool {
+	return false
```
True

rootfs commented:
I think `RequiresSELinuxRelabel` should always return `true` for volumes that need to be mounted. I am not sure if `git_repo` also needs relabel.

swagiaal commented:
> @swagiaal Here's a test we can do that doesn't rely on modifying any images:
> 
> Make a pod with MCS label c0,c1 that makes an emptyDir volume
> Make a pod with MCS label c2,c3 that runs on the host the pod in (1) is scheduled onto and does a hostpath mount of the volume directory for that pod.
> Verify that pod (2) cannot read or write pod (1)'s volume

Implemented here https://github.com/kubernetes/kubernetes/commit/506aaf149dca5efddcba1c35bc1abcc79b227a5d

ncdc commented:
@swagiaal rebase please? :smile:

ncdc commented on pkg/volume/volume.go:
```diff
@@ -45,6 +45,12 @@ type Builder interface {
+	// SupportsSELinux reports whether the given builder supports
+	// SELinux and would like the kubelet to relabel the volume to
+	// match the pod to which it will be attached. Usually this
+	// returns true of the volume type supports SELinux and the
```
"returns true of" -> "returns true **if**"

ncdc commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -586,13 +586,21 @@ func makeEnvList(envs []kubecontainer.EnvVar) (result []string) {
-func makeMountBindings(mounts []kubecontainer.Mount) (result []string) {
+// '<HostPath>:<ContainerPath>:ro', if the path is read only, or
+// '<HostPath>:<ContainerPath>:Z', if the volume requires SELinux
+// relabling and the pod provides an SELinux label
```
relabeling

ncdc commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -586,13 +586,21 @@ func makeEnvList(envs []kubecontainer.EnvVar) (result []string) {
 		if m.ReadOnly {
 			bind += ":ro"
 		}
+		// If the pod does not provide an SELinux context
```
This comment doesn't read correctly to me - would you consider revising it to be a bit clearer?

ncdc commented on test/e2e/security_context.go:
```diff
@@ -0,0 +1,214 @@
+		pod.Spec.Containers[0].Command = []string{"sleep", "6000"}
+
+		client := framework.Client.Pods(framework.Namespace.Name)
+		defer client.Delete(pod.Name, nil)
```
I'd move this to below the line that ensures there isn't an error creating the pod (just after 150)

ncdc commented on test/e2e/security_context.go:
```diff
@@ -0,0 +1,214 @@
+		pod.Spec.SecurityContext.SELinuxOptions = &api.SELinuxOptions{
+			Level: "s0:c2,c3",
+		}
+		defer client.Delete(pod.Name, nil)
```
Move to after line 205

swagiaal commented:
@pmorie @ncdc rebased and addressed feedback. I'll squash after the next round of the review.

pmorie commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -594,13 +594,24 @@ func makeEnvList(envs []kubecontainer.EnvVar) (result []string) {
 		if m.ReadOnly {
 			bind += ":ro"
 		}
+		// Only request relabling if the pod provides an
```
typo: relabeling

pmorie commented:
Nit: I would like the chcon runner in `pkg/util/chcon`, to follow the pattern of `pkg/util/mount`, `pkg/util/chmod`, etc.

swagiaal commented:
> So if I set an selinux context on both the pod and the container the volumes will match the pod's context and the container will run with the container context, correct?

That is correct.

> s/of/if

hmm... I think you are looking at an outdated commit.

> I noticed that volumes were being marked as labeled in the dockermanager. Do these two lines conflict? Is the dockermanager marking the volume as labeled prematurely?

the labeling in kubelet.go is to handle the hostIPC and hostPID cases. In those cases the kubelet does the labeling and sets the labeled boolean so the volume is skipped when asking docker to relabel. Please see function comment.

thockin commented:
Remove WIP in title?

swagiaal commented:
> I thought this was supposed to return true iff not being attached read-only?

@thockin that is handled here: https://github.com/kubernetes/kubernetes/pull/15323/files#diff-0db82891d463ba14dd59da9c77f4776eR644

but it is not handled for the hostIPC hostPID case.. I'll fix that

swagiaal commented:
> but it is not handled for the hostIPC hostPID case.. I'll fix that

@thockin Fixed https://github.com/kubernetes/kubernetes/commit/923af1726c5261f539c85d868b7cdef8e08c362b

thockin commented:
The mount being read-only is a VERY different thing than the device being
read-only.

Only the volume driver knows that - you have to ask the Build IsreadOnly()

On Mon, Oct 19, 2015 at 11:22 AM, Sami Wagiaalla <notifications@github.com>
wrote:

> I thought this was supposed to return true iff not being attached
> read-only?
>
> @thockin <https://github.com/thockin> that is handled here:
> https://github.com/kubernetes/kubernetes/pull/15323/files#diff-0db82891d463ba14dd59da9c77f4776eR644
>
> but it is not handled for the hostIPC hostPID case.. I'll fix that
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15323#issuecomment-149303957>
> .
>

swagiaal commented:
> Is this going to end up running once per mount or just once per volume?

Once per volume. I think that is enough though as SELinux labels are propagated by bind mount and and the label is a pod-level thing.

> The mount being read-only is a VERY different thing than the device being
> read-only.

Of course. My bad. This https://github.com/kubernetes/kubernetes/commit/988f7e8406a9c265510edc4d1d339cecf6e61093 address that. I also removed the check for the mount's IsReadOnly()

swagiaal commented:
> I also removed the check for the mount's IsReadOnly()

On second though maybe I should lave that in. If all the mounts are readOnly then perhaps we should not relabel even of the volume is not readOnly.

thockin commented:
I think it is correct to relabel if the volume is writeable, even if all pods want to access it as ro.

Can you rebase and squash this now that Paul's base change is in?  Then I'll review again

pmorie commented:
@thockin It's not merged yet -- it's waiting on shippable. :hourglass:

thockin commented:
aww hell, merge it tonight and rebase this for me tomorrow :)

On Wed, Oct 21, 2015 at 3:57 PM, Paul Morie <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> It's not merged yet -- it's waiting
> on shippable. [image: :hourglass:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15323#issuecomment-150047623>
> .
>

swagiaal commented:
> Can you rebase and squash this now that Paul's base change is in? Then I'll review again

@thockin rebased and squashed. PTAL

thockin commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -246,6 +246,10 @@ func (b *awsElasticBlockStoreBuilder) IsReadOnly() bool {
 }
 
+func (b *awsElasticBlockStoreBuilder) SupportsSELinux() bool {
+	return !b.IsReadOnly()
```
This needs to sync with https://github.com/kubernetes/kubernetes/pull/15352/files#r42777530 wherein I propose (reversing myself from previous, sorry) that volumes should not consider read-only-ness here, but the caller should.

pmorie replied:
I think I agree with @thockin, they're functionally equivalent but @thockin's way is less code.

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -955,6 +957,42 @@ func (kl *Kubelet) syncNodeStatus() {
+	volumeContext := fmt.Sprintf("%s:%s:%s:%s", rootDirSELinuxOptions.User, rootDirSELinuxOptions.Role, rootDirSELinuxOptions.Type, rootDirSELinuxOptions.Level)
+
+	for _, volume := range volumes {
+		if volume.Builder.SupportsSELinux() {
```
I thnk this should check builder.IsReadOnly() too

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -962,11 +1000,18 @@ func makeMounts(container *api.Container, podVolumes kubecontainer.VolumeMap) (m
 		}
+
+		relabelVolume := false
+		if vol.Builder.SupportsSELinux() && !vol.SELinuxLabeled && !mount.ReadOnly {
```
this should check builder.IsReadOnly() rather than mount.ReadOnly.

thockin commented on pkg/volume/empty_dir/empty_dir.go:
```diff
@@ -86,7 +87,7 @@ func (plugin *emptyDirPlugin) newBuilderInternal(spec *volume.Spec, pod *api.Pod
 		plugin:        plugin,
 		rootContext:   opts.RootContext,
-		chconRunner:   chconRunner,
+		chconRunner:   chconrunner,
```
Do we still need this to do it's own chcon at all?

pmorie replied:
I have the same question, @swagiaal.  I don't think emptyDir needs a chcon runner anymore.  It will need to pass the `rootcontext` argument to mount, but the kubelet can now do the chcon.

swagiaal commented:
@thockin I think all feedback is addressed now. PTAL

pmorie commented:
@swagiaal Rebase needed now that #15352 has merged.

swagiaal commented:
@pmorie, rebased

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -974,6 +976,42 @@ func (kl *Kubelet) syncNodeStatus() {
 
+// relabelVolumes relabels SELinux volumes to match the pod's
+// SELinuxOptions specification. This is only needed if the pod uses
+// hostPID or hostIPC. Otherwise relabeling is delegated to docker.
```
curiosity: if we feel like we know how to do the labelling ourselves, why should we ever defer to docker?  Isn't our code net simpler, less coupled, and more fixable if we just do it ourselves?

thockin replied:
e.g. what about non-docker runtimes?

swagiaal replied:
@thockin I think that we are better off letting docker, or non-docker rumtimes, do the labeling. Because docker is in a better position to know what label the container will run as if one is not provided. In this PR I use the rootContext as a heuristic to decide what the appropriate label 'type' will be. That works well because there are only two answers svirt_sandbox_file_t or "it does not matter because the container is privileged" but if docker changes its defaults we would have to handle that.

In addition I am investigating an issue where containers which run in the hostIPC or hostPID are unconfined by SELinux. That is true with pure docker but not in kube. If I can fix that issue this whole code pathway becomes unnecessary.

thockin replied:
I am not sure I buy that, but this PR is long overdue, so I'll stop arguing.

pmorie replied:
@thockin The chcon runner can definitely be re-used to provide support for rkt and other runtimes.  I'll make a follow-up issue.

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -974,6 +976,42 @@ func (kl *Kubelet) syncNodeStatus() {
+	for _, volume := range volumes {
+		if volume.Builder.SupportsSELinux() && !volume.Builder.IsReadOnly() {
+			dir := volume.Builder.GetPath()
+			err = chconRunner.SetContext(dir, volumeContext)
```
I don't know enough about chcon - does this need to be a recursive operation like chmod or just at the top-level dir?

thockin replied:
A comment could be nice.

swagiaal replied:
Ah yes.. I have fixed that

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -1079,6 +1127,16 @@ func (kl *Kubelet) GenerateRunContainerOptions(pod *api.Pod, container *api.Cont
+	// in the host pid or ipc namespaces so the kubelet must
+	// relabel the volumes
+	if pod.Spec.SecurityContext != nil && (pod.Spec.SecurityContext.HostIPC || pod.Spec.SecurityContext.HostPID) {
+		err = kl.relabelVolumes(pod, vol)
```
is it a problem that this whole function could fail and the labels are not reverted?  I think it is ok, but have to ask.

swagiaal replied:
@thockin With this PR it would not be too bad as the next pod will relabel the volume to match its label. So the volume is not rendered unusable. If we change things so that volumes are only relabeled once, only when newly formatted for example, then we might have to handle that. Although TBH I don't think there is value in putting effort into preserving the previous labels; default labels are not usable by pods, and presumably the user does not care about an older pod's labels because the want this pod to use the volume exclusively.

thockin commented:
LG overall, just a few questions

swagiaal commented:
> LG overall, just a few questions

@thockin I have addressed feedback PTAL. I'll squash before merging I just didn't want the responses above to get folded away

thockin commented:
LGTM

swagiaal commented:
@thockin all green. Could you re-add the LGTM tag. I squashed and rebased a few times to retrigger shippable

pmorie commented:
@bgrant0607 @thockin Can this get cherry-picked into 1.1?

thockin commented:
@brendandburns @ihmccreery @zmerlynn When is the next RC cut happening?

thockin commented:
you should submit the cherry pick ASAP - we can decide if that makes this cut or not

ihmccreery commented:
cc @quinton-hoole

I see neither a milestone nor a P0 label on this, so I'd advise against cherry picking.  We should only be merging things that are blockers.

quinton-hoole commented:
@ihmccreery , @thockin I have zero historical context here, but superficially this does indeed look like a very large, late change to go into 1.1.  And I don't see any e2e test coverage?  I'll leave @thockin to make the call, as he has the most context, and it falls within his area of expertise.

thockin commented:
Yeah, @pmorie - how bad do you really think it would be if this didn't go
into v1.1 at all?  How about if it went in for 1.1.2 or 1.1.3?

On Mon, Nov 2, 2015 at 3:02 PM, Quinton Hoole <notifications@github.com>
wrote:

> @ihmccreery <https://github.com/ihmccreery> , @thockin
> <https://github.com/thockin> I have zero historical context here, but
> superficially this does indeed look like a very large, late change to go
> into 1.1. And I don't see any e2e test coverage? I'll leave @thockin
> <https://github.com/thockin> to make the call, as he has the most
> context, and it falls within his area of expertise.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15323#issuecomment-153186099>
> .
>

pmorie commented:
@quinton-hoole @thockin There is e2e coverage for this and related PRs -- I would prefer to get them running in GCE before we pull them in for 1.1, so perhaps the timing isn't right.  I don't want to add a destabilizing feature at the last minute, so maybe the next minor release would be better.

Any opinion there, @smarterclayton ?

smarterclayton commented:
We're ok without it for 1.1 i think.

On Tue, Nov 3, 2015 at 11:43 AM, Paul Morie <notifications@github.com>
wrote:

> @quinton-hoole <https://github.com/quinton-hoole> @thockin
> <https://github.com/thockin> There is e2e coverage for this and related
> PRs -- I would prefer to get them running in GCE before we pull them in for
> 1.1, so perhaps the timing isn't right. I don't want to add a destabilizing
> feature at the last minute, so maybe the next minor release would be better.
>
> Any opinion there, @smarterclayton <https://github.com/smarterclayton> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15323#issuecomment-153411047>
> .
>
